src/utilities/contextual-code.ts

```diff
@@ -0,0 +1,168 @@
+
+import path = require('path');
+import * as vscode from 'vscode';
+import { SemanticTokensRegistrationType, SemanticTokensProviderShape, SymbolKind } from 'vscode-languageclient';
+import { ILspAnalyzer } from '../shared/types/LspAnalyzer';
+import { Outline } from '../shared/types/custom_protocols';
+import { getCodeForRange, isPositionInElementDefinitionRange } from '../shared/utils';
+import { Token } from '../shared/types/token';
+export class ContextualCodeProvider {
+
+    public async getContextualCode(document: vscode.TextDocument, range: vscode.Range, analyzer: ILspAnalyzer): Promise<string | undefined> {
+        const checkSymbols = (symbols: Outline[]): Outline | undefined => {
+            for (const symbol of symbols) {
+                const symbolRange = new vscode.Range(
+                    symbol.range.start.line,
+                    symbol.range.start.character,
+                    symbol.range.end.line,
+                    symbol.range.end.character,
+                );
+                if (isPositionInElementDefinitionRange(symbol, range.start)) {
+                    return symbol;
+                }
+                if (symbol.children) {
+                    const result = checkSymbols(symbol.children);
+                    if (result !== undefined) {
+                        return result;
+                    }
+                }
+            }
+            return undefined;
+        };
+        const outline = await analyzer.fileTracker.waitForOutline(document);
+
+        if (outline === undefined) {
+            return undefined;
+        }
+        const outlineSymbols = outline?.children || [];
+        // parse the outline to get all top level symbols
+        const symbol = checkSymbols(outlineSymbols);
+        if (symbol === undefined) {
+            return undefined;
+        }
+        const docTokens = await this.getDocumentTokens(document, analyzer, range);
+
+        const uniqueSymbols = docTokens
+            .filter(contextualSymbol => contextualSymbol.tokenType !== undefined && ["Class", "Method", "ENUM"].includes(contextualSymbol.tokenType))
+            .filter(contextualSymbol => contextualSymbol.name !== symbol.element.name)
+            .reduce((map, contextualSymbol) => {
+                if (contextualSymbol.code) {
+                    map.set(contextualSymbol.name, contextualSymbol.code);
+                }
+                return map;
+            }, new Map<string, string>());
+
+        let code: string = "";
+        for (const [key, value] of uniqueSymbols) {
+            const symbolCode = "```dart\n" + value + "\n```";
```

Comment:
- Append contextual code with the file path as well.
- Ensure that the same piece of contextual code is never duplicated. 
- Ensure all snippets belonging to the same file are grouped at once.